📖 Add clusterclass quickstart guide (#5374)
kubernetes-sigs/cluster-api: docs/book/src/tasks/experimental-features/cluster-classes.md

ClusterClass is currently behind a feature gate that needs to be enabled.
This tutorial will also use another experimental gated feature - Cluster Resource Set - in this tutorial. This is not required for ClusterClass but is used in this tutorial to set up networking.

To enable these features set the respective environment variables by running:
```bash
export EXP_CLUSTER_RESOURCE_SET=true
export CLUSTER_TOPOLOGY=true
```
This ensures that the Cluster Topology and Cluster Resource Set features are enabled when the providers are initialized.

#### Create a CAPI management cluster

A script to set up a Kind cluster pre-configured for CAPD (the docker infrastructure provider) can be found in the [hack folder of the core repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/hack/kind-install-for-capd.sh). 

To set up the cluster from the root of the repo run:
```bash
./hack/kind-install-for-capd.sh
clusterctl init --infrastructure docker
````

### Create a new Cluster using ClusterClass

#### Create the ClusterClass and templates

With a management Cluster with CAPD and Cluster Topology the ClusterClass and other templates needed to stamp clusters can be created.

> The below contains six yaml objects which are used to define the cluster:

Reviewer: ```suggestion
The ClusterClass - first in the below yaml - contains references to the templates needed to build a full cluster, thus defining a shape that can be re-used as many times as you want.
```

* ClusterClass
* For the InfrastructureCluster:
  * DockerClusterTemplate
* For the ControlPlane:
  * KubeadmControlPlaneTemplate
  * DockerMachineTemplate 
* For the worker nodes:
  * DockerMachineTemplate
  * KubeadmConfigTemplate

> The ClusterClass - first in the below yaml - contains references to the templates needed to build a full cluster using the shape defined in the ClusterClass.

Reviewer: ```suggestion
```

> The full ClusterClass definition can also be found in the [CAPI repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/docs/book/src/tasks/yamls/clusterclass.yaml).

Reviewer: this is a 404 🤔

Reviewer 2: Until the PR is merged? (hopefully :))

<details><summary>ClusterClass</summary>

```yaml
{{#include ./yamls/clusterclass.yaml}}
```

</details>

To create the objects on your local cluster run:

```bash

> kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/docs/book/src/tasks/yamls/clusterclass.yaml 

Reviewer: this is a 404 🤔

```

#### Enable networking for workload clusters

To make sure workload clusters come up with a functioning network a Kindnet ConfigMap with a Kindnet ClusterResourceSet is required. Kindnet only offers networking for Clusters built with Kind and CAPD. This can be substituted for any other networking solution for Kubernetes e.g. Calico as used in the Quickstart guide.

> The kindnet configuration file can be found in the [CAPI repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/docs/book/src/tasks/yamls/clusterclass.yaml).

Reviewer: this is a 404 🤔

kubeadmcontrolplane.controlplane.cluster.x-k8s.io/clusterclass-quickstart-XXXX    clusterclass-quickstart   true          true                   1          1       1         0             7m29s   v1.22.0

NAME                                                                             CLUSTER                   REPLICAS   READY   UPDATED   UNAVAILABLE   PHASE     AGE     VERSION
machinedeployment.cluster.x-k8s.io/clusterclass-quickstart-linux-workers-XXXX    clusterclass-quickstart   1          1       1         0             Running   7m29s   v1.22.0
```

## Clean Up

Delete workload cluster.
```bash
kubectl delete cluster clusterclass-quickstart
```
<aside class="note warning">

IMPORTANT: In order to ensure a proper cleanup of your infrastructure you must always delete the cluster object. Deleting the entire cluster template with `kubectl delete -f clusterclass-quickstart.yaml` might lead to pending resources to be cleaned up manually.
</aside>

Delete management cluster
```bash
kind delete clusters capi-test
```

<!-- links -->
[quick start guide]: ../../user/quick-start.md
[bootstrap cluster]: ../../reference/glossary.md#bootstrap-cluster

> [clusterctl]: ../../clusterctl/overview.md

Reviewer: This is the clusterctl overview, but the users can't find instruction on how to install it here. should we point to quick-start.html#install-clusterctl instead?

    workers:
      machineDeployments:
        - class: linux-worker
          name: linux-workers
          replicas: 1 
```

Create the Cluster object from the file in [the CAPI repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/clusterclass-quickstart.yaml) with:

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/clusterclass-quickstart.yaml
```

#### Verify the workload cluster is running

The cluster will now start provisioning. You can check status with:

```bash
kubectl get cluster
```

Get a view of the cluster and its resources as they are created by running:

```bash
clusterctl describe cluster clusterclass-quickstart
```

To verify the control plane is up:

```bash
kubectl get kubeadmcontrolplane
```

The output should be similar to:

```bash
NAME                            INITIALIZED   API SERVER AVAILABLE   VERSION   REPLICAS   READY   UPDATED   UNAVAILABLE
clusterclass-quickstart                 true                                 v1.21.2   1                  1         1
```

## Upgrade a Cluster using Managed Topology

> The `Spec.Topology` field added to the Cluster object as part of ClusterClass allows changes made on the Cluster to be propagated across all relevant objects. This turns a Kubernetes cluster upgrade into a one-touch operation.

Reviewer 2: ```suggestion
The `spec.topology` field added to the Cluster object as part of ClusterClass allows changes made on the Cluster to be propagated across all relevant objects. This turns a Kubernetes cluster upgrade into a one-touch operation.
```
nit: to align to the usage below (or the other way around)

Reviewer 2: @killianmuldoon can you also "lowercase" the `topology` part of it, pls?

Looking at the newly-created cluster, the version of the control plane and the machine deployments is v1.21.2.

```bash
> kubectl get kubeadmcontrolplane,machinedeployments

NAME                                                                      CLUSTER           INITIALIZED   API SERVER AVAILABLE   REPLICAS   READY   UPDATED   UNAVAILABLE   AGE     VERSION

> kubeadmcontrolplane.controlplane.cluster.x-k8s.io/clusterclass-quickstart-XXXX    clusterclass-quickstart   true          true                   1          1       1         0             2m21s   v1.21.2

Reviewer 2: The columns seem to be a bit off (at least in the rendered book) (also in other cases)

Author: Copy paste issue

# ClusterClass

The ClusterClass feature introduces a new way to create clusters which reduces boilerplate and enables flexible and powerful customization of clusters.
ClusterClass is a powerful abstraction implemented on top of existing interfaces and offering a set of tools and operations to streamline cluster lifecycle management while maintaining the same underlying API.

This tutorial covers using Cluster API and ClusterClass to create a Kubernetes cluster and how to perform a one-touch Kubernetes version upgrade.

<aside class="note warning">

<h1>Note</h1>

> This guide covers creating a ClusterClass cluster with the CAPD provider.

Author: The CAPD provider is the key part here though - I'd like to really highlight what this guide does - and hopefully we can expand it for other providers sooner rather than later.

</aside>

## Installation

### Prerequisites

#### Install tools

This guide requires the following tools are installed: 
- Install and setup [kubectl] and [clusterctl]
- Install [Kind] and [Docker]

> For more information see the [quick start guide].

Reviewer 2: Up to you but I would probably drop this sentence. As a reader I would be confused at which part of the Quickstart to look at for more information as I don't know which part also applies to ClusterClass and which doesn't.

Are we referencing to a specific section of the Quickstart? (then we could maybe make the link more specific)

Author: I think you're right - better to make this guide stand by itself.

#### Enable experimental features

ClusterClass is currently behind a feature gate that needs to be enabled.
This tutorial will also use another experimental gated feature - Cluster Resource Set - in this tutorial. This is not required for ClusterClass but is used in this tutorial to set up networking.

> To enable these features set the respective environmental variables by running:

Reviewer 2: ```suggestion
To enable these features set the respective environment variables by running:
```
nit (but just my preference, as that's what they are usually called in my experience)

```bash
export EXP_CLUSTER_RESOURCE_SET=true
export CLUSTER_TOPOLOGY=true
```
This ensures that the Cluster Topology and Cluster Resource Set features are enabled when we initialize providers.

#### Create a CAPI management cluster

A script to set up a Kind cluster pre-configured for CAPD ( the docker infrastructure provider) can be found in the [hack folder of the core repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/hack/kind-install-for-capd.sh). 

To set up the cluster from the root of the repo run:
```bash
./hack/kind-install-for-capd.sh
clusterctl init --infrastructure docker
````

### Create a new Cluster using ClusterClass

#### Create the ClusterClass and templates

Now that we have a management Cluster with CAPD and Cluster Topology enabled we can create the ClusterClass and the templates needed to stamp clusters.

The below contains six yaml objects which are used to define the cluster:

> *ClusterClass

Reviewer 2: ```suggestion
* ClusterClass
```
The missing space leads to:
![Screenshot 2021-10-05 at 09 04 04](https://user-images.githubusercontent.com/4662360/135975802-0e860ab6-e2f7-45b2-befa-c886d7d0088e.png)

* For the InfrastructureCluster:
  * DockerClusterTemplate

> * For the Control Plane:

Reviewer 2: ```suggestion
* For the ControlPlane:
```
for consistency with `InfrastructureCluster` (or the other way around)

        joinConfiguration:
          nodeRegistration:
            criSocket: /var/run/containerd/containerd.sock
            kubeletExtraArgs:
              cgroup-driver: cgroupfs
              eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'
      version: v1.21.2
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: DockerMachineTemplate
metadata:
  name: "clusterclass-control-plane"
  namespace: default
spec:
  template:
    spec:
      extraMounts:
        - containerPath: "/var/run/docker.sock"
          hostPath: "/var/run/docker.sock"
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: DockerMachineTemplate
metadata:
  name: "clusterclass-md-1"
  namespace: default
spec:
  template:
    spec: { }
---
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
metadata:
  name: "clusterclass-md-1"
  namespace: default
spec:
  template:
    spec:
      joinConfiguration:
        nodeRegistration:
          kubeletExtraArgs:
            cgroup-driver: cgroupfs
            eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'
```

> The above file can be found in the [CAPI repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/clusterclass.yaml).

Reviewer 2: You could drop this sentence if you want as the information is already part of the kubectl apply command below (same for the kindnet CRS file)

Author: I think the explicitness might help users find the docs here.

To create the objects on your local cluster run:

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/clusterclass.yaml 
```

#### Enable networking for our clusters

To make sure our workload clusters come up with a functioning network we can define a Kindnet ConfigMap with a Kindnet ClusterResourceSet. Kindnet only offers networking for Clusters built with Kind and CAPD. This can be substituted for any other networking solution for Kubernetes e.g. Calico as used in the Quickstart guide.

> `

Reviewer 2: ```suggestion

```

The kindnet configuration file can be found in the [CAPI repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/kindnet-clusterresourceset.yaml).

To create the resources run:
```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/kindnet-clusterresourceset.yaml 
```

> #### Apply the workload cluster

Reviewer 2: ```suggestion
#### Create the workload cluster
```
maybe?

    serviceDomain: "cluster.local"
  topology:
    class: clusterclass
    version: v1.21.2
    controlPlane:
      replicas: 1 
    workers:
      machineDeployments:
        - class: linux-worker
          name: linux-workers
          replicas: 1 
```

We can create the Cluster object from the file in [the CAPI repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/clusterclass-quickstart.yaml) with:

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/test/infrastructure/docker/examples/clusterclass-quickstart.yaml
```

#### Verify the workload cluster is running

The cluster will now start provisioning. You can check status with:

```bash
kubectl get cluster
```

You can get a view of the cluster and its resources as they are created by running:

```bash
clusterctl describe cluster clusterclass-quickstart
```

To verify the control plane is up:

```bash
kubectl get kubeadmcontrolplane
```

You should see an output similar to the below:

```bash
NAME                            INITIALIZED   API SERVER AVAILABLE   VERSION   REPLICAS   READY   UPDATED   UNAVAILABLE

> capi-quickstart                 true                                 v1.21.2   1                  1         1

Reviewer 2: I still don't really like `clusterclass-quickstart` as all resources are now prefixed with `clusterclass` (maybe `capi-clusterclass-quickstart` would be a compromise?) (feel free to ignore this comment)

But apart from that let's make the doc consistent. Please replace all occurrences of `capi-quickstart` with the cluster name.

Author: Updated the outstanding references. I don't really think capi-clusterclass adds anything over what clusterclass does, but if someone has strong feelings on the naming I've no problem changing it.

```

## Upgrade a Cluster using Managed Topology

The `Spec.Topology` field added to the Cluster object as part of ClusterClass allows changes made on the Cluster to be propogated across all relevant objects. This turns a Kubernetes cluster upgrade into a one-touch operation.
Looking at the cluster we just created we can see the version of the control plane and the other machines is v1.21.2.

```bash
> kubectl get kubeadmcontrolplane,machinedeployments

> 
> 

Reviewer 2: ```suggestion
```
nit: Maybe drop those empty lines. It's a lot of empty space in the rendered output.

NAME                                                                      CLUSTER           INITIALIZED   API SERVER AVAILABLE   REPLICAS   READY   UPDATED   UNAVAILABLE   AGE     VERSION
kubeadmcontrolplane.controlplane.cluster.x-k8s.io/capi-quickstart-XXXX    capi-quickstart   true          true                   1          1       1         0             2m21s   v1.21.2

NAME                                                                     CLUSTER           REPLICAS   READY   UPDATED   UNAVAILABLE   PHASE     AGE     VERSION
machinedeployment.cluster.x-k8s.io/capi-quickstart-linux-workers-XXXX    capi-quickstart   1          1       1         0             Running   2m21s   v1.21.2

```

> To update the Cluster the only change needed is to the 'version' field under `spec.topology` in the Cluster object.

Reviewer 2: ```suggestion
To update the Cluster the only change needed is to the `version` field under `spec.topology` in the Cluster object.
```
nit: maybe use them instead of `'` then it's also rendered as code in the book

> Change '1.21.2' to '1.22.0' as below. 

Reviewer 2: ```suggestion
Change `v1.21.2` to `v1.22.0` as below. 
```

> # Cluster Class

Reviewer 3: ```suggestion
# ClusterClass
```

> The ClusterClass feature introduces a new way to create clusters which reduces boilerplate and enables flexible and powerful customization of CAPI-created clusters.

Reviewer 3: ```suggestion
The ClusterClass feature introduces a new way to create clusters which reduces boilerplate and enables flexible and powerful customization of clusters.
```

> ClusterClass is implemented on top of existing CAPI interfaces and offers a set of tools and operations to streamline cluster lifecycle management while maintaining the same underlying API.

Reviewer 3: ```suggestion
ClusterClass is a powerful abstraction implemented on top of existing interfaces offering a set of tools and operations to streamline cluster lifecycle management while maintaining the same underlying API.
```

> This tutorial covers using Cluster API and Cluster Class to create a Kubernetes cluster and how to perform a one-touch Kubernetes update procedure using ClusterClass features.

Reviewer 3: ```suggestion
This tutorial covers using Cluster API and ClusterClass to create a Kubernetes cluster and how to perform a one-touch Kubernetes version upgrade.
```

<aside class="note warning">

<h1>Note</h1>

This guide covers creating a ClusterClass cluster with the CAPD provider.
</aside>

## Installation

> ### Common Prerequisites
> This guide is based on having the same pre-requisites as the CAPI [quick start guide].

Reviewer 3: ```suggestion
### Common Prerequisites

This guide is based on having the same pre-requisites as the CAPI [quick start guide].
```

> That is:

Reviewer 3: ```suggestion
```

> - Install and setup [kubectl]  and [clusterctl]

Reviewer 3: ```suggestion
- Install and setup [kubectl] and [clusterctl]
```

- Install [Kind] and [Docker]

> ClusterClass is currently behind a feature gate that needs to be enabled.

Reviewer 3: Add this as a pre-requisite?

Author: This is all under pre-requisites, but I'll break it up into tool installation and then the commands.

> This tutorial will also use another experimental gated feature - Cluster Resource Set - in this tutorial. This is not required for ClusterClass but is used in this tutorial to set up networking.

Reviewer: ```suggestion
This tutorial will also use another experimental gated feature - Cluster Resource Set -. This is not required for ClusterClass to work, but is used in this tutorial to set up networking.
```

To enable these features set the respective environmental variables by running:
```bash
export EXP_CLUSTER_RESOURCE_SET=true
export CLUSTER_TOPOLOGY=true
```

> This ensures that the Cluster Topology and Cluster Resource Set features are enabled when we initialize our CAPI Providers.

Reviewer 3: ```suggestion
This ensures that the Cluster Topology and Cluster Resource Set features are enabled when we initialize providers.
```

> A script to set up a Kind cluster pre-configured for CAPD can be found in the [hack folder of the core repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/hack/kind-install-for-capd.sh). 

Reviewer 3: ```suggestion
A script to set up a Kind cluster pre-configured for CAPD (docker infrastructure provider) can be found in the [hack folder of the core repo](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/main/hack/kind-install-for-capd.sh). 
```

To set up the cluster from the root of the repo run:
```bash
./hack/kind-install-for-capd.sh
clusterctl init --infrastructure docker
````

#### Creating the ClusterClass and templates

> Now that we have a CAPI Management Cluster with CAPD and Cluster Topology enabled we can create the ClusterClass and the templates needed to stamp clusters.

Reviewer 3: ```suggestion
Now that we have a management Cluster with CAPD and Cluster Topology enabled we can create the ClusterClass and the templates needed to stamp clusters.
```

> #### Creating the ClusterClass and templates
> Now that we have a CAPI Management Cluster with CAPD and Cluster Topology enabled we can create the ClusterClass and the templates needed to stamp clusters.

Reviewer 3: ```suggestion
#### Creating the ClusterClass and templates

Now that we have a CAPI Management Cluster with CAPD and Cluster Topology enabled we can create the ClusterClass and the templates needed to stamp clusters.
```

            kubeletExtraArgs:
              cgroup-driver: cgroupfs
              eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'
      version: v1.21.2
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: DockerMachineTemplate
metadata:
  name: "my-clusterclass-control-plane"
  namespace: default
spec:
  template:
    spec:
      extraMounts:
        - containerPath: "/var/run/docker.sock"
          hostPath: "/var/run/docker.sock"
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: DockerMachineTemplate
metadata:
  name: "my-clusterclass-md-1"
  namespace: default
spec:
  template:
    spec: { }
---
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
metadata:
  name: "my-clusterclass-md-0"
  namespace: default
spec:
  template:
    spec:
      joinConfiguration:
        nodeRegistration:
          kubeletExtraArgs:
            cgroup-driver: cgroupfs
            eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'
```

Copy the above yaml into a local file - clusterclass.yaml - and run:

```bash
kubectl apply -f clusterclass.yaml
```

> #### Enable networking for our clusters
> To make sure our workload clusters come up with a functioning network we can define a Kindnet ConfigMap with a Kindnet ClusterResourceSet. Kindnet only offers networking for Clusters built with Kind and CAPD. This can be substituted for any other networking solution for Kubernetes e.g. Calico as used in the Quickstart guide.

Reviewer 3: ```suggestion
#### Enable networking for our clusters

To make sure our workload clusters come up with a functioning network we can define a Kindnet ConfigMap with a Kindnet ClusterResourceSet. Kindnet only offers networking for Clusters built with Kind and CAPD. This can be substituted for any other networking solution for Kubernetes e.g. Calico as used in the Quickstart guide.
```

                  readOnly: false
                - name: lib-modules
                  mountPath: /lib/modules
                  readOnly: true
              resources:
                requests:
                  cpu: "100m"
                  memory: "50Mi"
                limits:
                  cpu: "100m"
                  memory: "50Mi"
              securityContext:
                privileged: false
                capabilities:
                  add: ["NET_RAW", "NET_ADMIN"]
          volumes:
            - name: cni-cfg
              hostPath:
                path: /etc/cni/net.d
            - name: xtables-lock
              hostPath:
                path: /run/xtables.lock
                type: FileOrCreate
            - name: lib-modules
              hostPath:
                path: /lib/modules
---
apiVersion: addons.cluster.x-k8s.io/v1beta1
kind: ClusterResourceSet
metadata:
  name: cni
spec:
  strategy: ApplyOnce
  clusterSelector:
    matchLabels:
      cni: kindnet
  resources:
    - name: kindnet
      kind: ConfigMap
```

Copy the above yaml into a local text file - kindnet.yaml - and run:
```bash
kubectl apply -f kindnet.yaml
```

> #### Apply the workload cluster
> The below is a Cluster definition that leverages the ClusterClass created above to define its shape.

Reviewer 3: ```suggestion
#### Apply the workload cluster

The below is a Cluster definition that leverages the ClusterClass created above to define its shape.
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:

>   name: "capi-quickstart"

Reviewer 3: Should we call this something else? Isn't `capi-quickstart` the same Cluster that's in today's quickstart?

  namespace: default
  labels:
    cni: kindnet
spec:
  clusterNetwork:
    services:
      cidrBlocks: ["10.128.0.0/12"]
    pods:
      cidrBlocks: ["192.168.0.0/16"]
    serviceDomain: "cluster.local"
  topology:

>     class: my-clusterclass

Author: I'll change the references back to just 'clusterclass'

> # Cluster Class

Reviewer 2: Can you reference this file in the respective section in SUMMARY.md? Otherwise the doc is not easily accessible in the book.

Reviewer 2: Also to `experimental-features.md`

The ClusterClass feature introduces a new way to create clusters which reduces boilerplate required and allows flexible and powerful customization of CAPI-created clusters.

> ClusterClass is implemented on top of pre-existing CAPI interfaces and will 

Reviewer 2: There seems to be something missing at the end of the sentence

This tutorial covers using Cluster API and Cluster Class to create a Kubernetes cluster and how to perform a one-touch Kubernetes update procedure using ClusterClass features.

<aside class="note warning">

<h1>Note</h1>

This guide covers creating a ClusterClass cluster with the CAPD provider.
</aside>

## Installation

### Common Prerequisites
This guide is based on having the same pre-requisites as the CAPI [quick start guide].
That is:
- Install and setup [kubectl]  and [clusterctl]
- Install [Kind] and [Docker]

> ClusterClass is currently behind a feature gate that needs to be enabled. To do so run:

Reviewer: Let's explicity say that in this tutorial we also use another experimental feature, CRS, so we need to enable both
(I saw it is down in the paragraph, but we need something before the export)

```bash

> export  EXP_CLUSTER_RESOURCE_SET=true

Reviewer 2: ```suggestion
export EXP_CLUSTER_RESOURCE_SET=true
```
nit (double-space just looks a bit strange in the rendered view)

> export CLUSTER_TOPOLOGY=true

Reviewer 2: I wonder if the env var should actually have been `EXP_CLUSTER_TOPOLOGY` like all our other feature flag env vars.

Author: Yeah - I guess it's not all that important, but there's definitely a feature flag drift here - but ClusterClass is also outside the EXP directory so it is implemented in a different way to start.

Reviewer 2: Good point. Okay no idea how it should be :) (@fabriziopandini ?)

Reviewer 3: We should probably not change the feature gate at this point. Would we see instead that new feature gates shouldn't have the EXP prefix instead, given that are all by definition experiments?

Reviewer 2: I agree that the FF prefix isn't really helpful and that we shouldn't change anything now. Usually I think it's a good idea to have some namespacing in env variables (e.g. CAPI_FF prefix), but I think it's too late for that. 3 different formats for this are too much :)

```

> This ensures that Cluster Topology and Cluster Resource Set are enabled when we initialize our CAPI Providers.

Reviewer 2: ```suggestion
This ensures that the Cluster Topology and Cluster Resource Set features are enabled when we initialize our CAPI Providers.
```

> *Note*: Cluster Resource Set is not required for ClusterClass but is used in this tutorial to set up networking.

Reviewer 2: ```suggestion

*Note*: Cluster Resource Set is not required for ClusterClass but is used in this tutorial to set up networking.
```
Looks like this needs an additional new line, otherwise it will be rendered without line break in the same line as the sentence above.

Fyi, not sure if you're aware. You can view the book locally (with auto-refresh) via:
```bash
cd  ./docs/books
make serve
# http://localhost:3000/tasks/experimental-features/cluster-classes.html
# I think the link only works after adding the file to SUMMARY.md)
```

Author: I didn't really think putting the Note on its own line was needed here, but I'll update.

Reviewer 2: Ah okay, just thought that was intended as it was already on a new line in the markdown

A script to set up a Kind cluster pre-configured for CAPD can be found in the hack folder. To set up the cluster run:
```bash
./hack/kind-install-for-capd.sh
clusterctl init --infrastructure docker
````

Alternatively the quickstart guide for setting up a CAPD enabled cluster can be followed.

#### Creating the ClusterClass and templates
Now that we have a CAPI Management Cluster with CAPD and Cluster Topology enabled we can create the ClusterClass and the templates needed to stamp clusters.

The below contains six yaml objects which are used to define the cluster:

> 1) ClusterClass
> 1) DockerMachineTemplate
> 1) DockerClusterTemplate
> 1) DockerMachineTemplate for ControlPlane
> 1) KubeadmControlPlaneTemplate
> 1) KubeadmConfigTemplate

Reviewer 2: ```suggestion
* ClusterClass
* For the control plane:
  * KubeadmControlPlaneTemplate
  * DockerMachineTemplate
* For the infrastructure cluster:
  * DockerClusterTemplate
* For the worker nodes:
  * KubeadmConfigTemplate
  * DockerMachineTemplate
```
Maybe a bit better because it shows the structure, but up to you.

Reviewer 2: Might be good to sort the resources in the YAML in the same order

The ClusterClass - first in the below yaml - contains references to the templates needed to build a full cluster using the shape defined in the ClusterClass.

> ```yaml

Reviewer 2: Just an idea. Should we move this file and the next to: ./test/infrastructure/docker/examples and then just provide a kubectl apply command? Would make the quickstart way more compact.

Author: @fabriziopandini what do you think? I know the clusterctl generate flow is on the way, but it would make the guide more compact and less error prone I think.

Reviewer: I'm ok with both solutions. Up to you

Author: Wondering what the best solution is here - I've left the clusterclass templates and the clusteryaml in the guide so that we're explicit about the objects that users are working with, but I've moved them under the test/infrastructure package so that they can be easily created with kubectl apply - any thoughts on the current form? I think showing the yamls in the tutorial has some value. Do we have any mechanism for embedding yamls in the book?

Reviewer 2: I'm fine with the current version. Not sure if there's a way to wrap it in something which can be expanded/collapsed on demand (I couldn't find an example in our book)

Reviewer 2: @killianmuldoon I think you can try `details` if you want, it's used in the kubebuilder book:
* https://book.kubebuilder.io/cronjob-tutorial/basic-project.html
* https://raw.githubusercontent.com/kubernetes-sigs/kubebuilder/master/docs/book/src/cronjob-tutorial/basic-project.md

Looks like a good compromise to me.

Reviewer 2: Ah now I got what you mean with embedding. Afaik you can use things like `{{#include ./testdata/project/PROJECT}}` (also shown in the kubebuilder example I linked)

Author: I've done this and moved the yamls to be inside the book folder rather than in examples. This might be more appropriate for now - especially if those were going to be removed if clusterctl generate replaces raw yamls here. WDYT?

Reviewer 2: I think you're right.

> apiVersion: cluster.x-k8s.io/v1alpha4

Reviewer 2: Please update the apiVersions to v1beta1

kind: ClusterClass
metadata:
  name: cluster-class
  namespace: default
spec:
  controlPlane:
    metadata:
    ref:
      apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
      kind: KubeadmControlPlaneTemplate

>       name: my-cluster-control-plane

Reviewer 2: I wonder if we should call the ClusterClass `my-cluster-class` and then all the referenced templates `my-cluster-class-<suffix>`. The current names implicate a bit that they are for my-cluster and not for my-cluster-class

      namespace: default
    machineInfrastructure:
      ref:
        kind: DockerMachineTemplate
        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
        name: "my-cluster-control-plane"
        namespace: default
  infrastructure:
    ref:
      apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
      kind: DockerClusterTemplate
      name: my-cluster
      namespace: default
  workers:
    machineDeployments:
    - class: linux-worker
      template:
        metadata:
        bootstrap:
          ref:
            kind: KubeadmConfigTemplate
            apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
            name: "my-cluster-md-0"
        infrastructure:
          ref:
            kind: DockerMachineTemplate
            apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4

>             name: "md1"

Reviewer 2: Should we call this the same as the bootstrap config template? md-0 vs m1 looks not ideal.

                  cpu: "100m"
                  memory: "50Mi"
                limits:
                  cpu: "100m"
                  memory: "50Mi"
              securityContext:
                privileged: false
                capabilities:
                  add: ["NET_RAW", "NET_ADMIN"]
          volumes:
            - name: cni-cfg
              hostPath:
                path: /etc/cni/net.d
            - name: xtables-lock
              hostPath:
                path: /run/xtables.lock
                type: FileOrCreate
            - name: lib-modules
              hostPath:
                path: /lib/modules
---
apiVersion: addons.cluster.x-k8s.io/v1alpha4
kind: ClusterResourceSet
metadata:
  name: cni
spec:
  strategy: ApplyOnce
  clusterSelector:
    matchLabels:
      cni: kindnet
  resources:
    - name: kindnet
      kind: ConfigMap
```

Copy the above yaml into a local text file - kindnet.yaml - and run:
```bash
kubectl apply -f kindnet.yaml
```

#### Apply the workload cluster
The below is a Cluster definition that leverages the ClusterClass created above to define its shape.
```yaml
apiVersion: cluster.x-k8s.io/v1alpha4
kind: Cluster
metadata:

>   name: "clusterclass-quickstart"

Reviewer 2: nit: maybe we should not calls this cluster `clusterclass-<suffix>` (to avoid confusion with clusterclass especially in the generated topology resources). Maybe just `capi-quickstart` like in the regular quickstart?

  namespace: default
  labels:
    cni: kindnet

> spec:
>   clusterNetwork:
>     services:
>       cidrBlocks: ["10.128.0.0/12"]
>     pods:
>       cidrBlocks: ["192.168.0.0/16"]
>     serviceDomain: "cluster.local"
>   topology:
>     class: cluster-class
>     version: v1.21.2
>     controlPlane:
>       metadata:
>         labels:
>           topologyLabel: topologyLabelValue
>         annotations:
>           topologyAnnotation: topologyAnnotation
>       replicas: 1 
>     workers:
>       machineDeployments:
>         - class: linux-worker
>           name: linux-workers
>           metadata:
>             labels:
>               topologyLabel: topologyLabelValue
>             annotations:
>               topologyAnnotation: topologyAnnotation
>           replicas: 1 

Reviewer 2: ```suggestion
spec:
  topology:
    class: cluster-class
    version: v1.21.2
    controlPlane:
      replicas: 1 
    workers:
      machineDeployments:
        - class: linux-worker
          name: linux-workers
          replicas: 1 
```

Should we shorten the YAML to this? Imho it shows better how simple it is to create a Cluster with ClusterClass

Author: Looks like the simplest thing to do is to keep the kindnet config but get rid of the labels.

Author: Kindnet needs the values in the clusterNetwork it seems or else the nodes can't come up - is there somewhere else I can set them to get around this?

Reviewer 2: I have no idea how kindnet is configured :). But I think it's fine to leave it when it's required by kindnet.

```

To create the cluster save the above yaml in a local file - clusterclass-quickstart.yaml - and run:

```bash
kubectl apply -f clusterclass-quickstart.yaml
```
#### Accessing the workload cluster

The cluster will now start provisioning. You can check status with:

```bash
kubectl get cluster
```

You can get a view of the cluster and its resources as they are created by running:

```bash
clusterctl describe cluster clusterclass-quickstart
```

To verify the control plane is up:

```bash
kubectl get kubeadmcontrolplane
```

You should see an output similar to the below:

```bash
NAME                            INITIALIZED   API SERVER AVAILABLE   VERSION   REPLICAS   READY   UPDATED   UNAVAILABLE
clusterclass-quickstart         true                                 v1.21.2   1                  1         1
```

### Upgrade a Cluster using Managed Topology

> The Spec.Topology field added to the Cluster object as part of ClusterClass allows changes made on the Cluster to be propogated across all relevant objects. This turns a Kubernetes cluster upgrade into a one touch operation.

Reviewer 2: ```suggestion
The `.spec.topology` field added to the Cluster object as part of ClusterClass allows changes made on the Cluster to be propagated across all relevant objects. This turns a Kubernetes cluster upgrade into a one-touch operation.
```
I'm not sure how to write one-touch. But that's how it was written above.

Looking at the cluster we just created we can see the version of the control plane and the other machines is v1.21.2.

```bash

> > kubectl get kubeadmcontrolplane,machines

Reviewer: what about getting machine deployment as well, so the user can appreciate that upgrades impact both?

Author: I'm using machines because machinedeployment doesn't have a version field in its kubectl get output - I think it's fairly clear to the user, but we could try to include a version in the machinedeployment columns or else add some text explaining what the machines are. WDYT?

Reviewer 2: It should be there, otherwise it's a bug: https://github.com/kubernetes-sigs/cluster-api/blob/07c0a4809361927b15cde2747b34142b7c7ead15/api/v1beta1/machinedeployment_types.go#L285

NAME                            INITIALIZED   API SERVER AVAILABLE   VERSION   REPLICAS   READY   UPDATED   UNAVAILABLE
clusterclass-quickstart         true          true                   v1.21.2   1                  1         1

NAME                                                           CLUSTER                   AGE     PROVIDERID                                                                PHASE     VERSION

> clusterclass-quickstart-linux-workers-XXXX                     clusterclass-quickstart   2m51s   docker:////clusterclass-quickstart-linux-workers-XXXX                    Running   v1.21.2

Reviewer 2: I think it's fine to keep the real machine names in the output (makes it a bit more realistic, too)

Author: I think it's better to clean up the hashes like this - unless there's a good reason to keep them I think this is more readable.

Reviewer 2: Okay, up to you.

clusterclass-quickstart-XXXX                                   clusterclass-quickstart   2m48s   docker:////clusterclass-quickstart-XXXX                                  Running   v1.21.2

```

> To update the Cluster the only change needed is to the 'version' field under spec.topology in the Cluster object. Change '1.21.2' to '1.22' as below. 

Reviewer 2: ```suggestion
To update the Cluster the only change needed is to the 'version' field under `spec.topology` in the Cluster object. Change '1.21.2' to '1.22.0' as below. 
```

```yaml
apiVersion: cluster.x-k8s.io/v1alpha4
kind: Cluster
metadata:
  name: "clusterclass-quickstart"
  namespace: default
  labels:
    cni: kindnet
spec:
  clusterNetwork:
    services:
      cidrBlocks: ["10.128.0.0/12"]
    pods:
      cidrBlocks: ["192.168.0.0/16"]
    serviceDomain: "cluster.local"
  topology:
    class: cluster-class
    version: v1.22.0
      controlPlane:
      metadata:
        labels:
          topologyLabel: topologyLabelValue
        annotations:
          topologyAnnotation: topologyAnnotation
      replicas: 1 
    workers:
      machineDeployments:
        - class: linux-worker
          name: linux-workers
          metadata:
            labels:
              topologyLabel: topologyLabelValue
            annotations:
              topologyAnnotation: topologyAnnotation
          replicas: 1 
```

Save the above change to your local file and apply the update to the Cluster with:
```bash

> kubectl apply -f clusterclass.yaml

Reviewer 2: Can we make this a one-line kubectl patch command?

```

> The upgrade will take a minute to roll out as it will take place machine by machine with older versions of the machines only being removed after health newer versions come online.

Reviewer 2: ```suggestion
The upgrade will take a bit to roll out as it will take place machine by machine with older versions of the machines only being removed after healthy newer versions come online.
```
(not sure about "a minute" :))

To watch the update progress run:
```bash

> watch kubectl get  kubeadmcontrolplane,machines

Reviewer 2: ```suggestion
watch kubectl get kubeadmcontrolplane,machines
```
nit

```
After a few minutes the upgrade will be complete and the output will be similar to:
```bash
NAME                                                                              AGE   INITIALIZED   API SERVER AVAILABLE   VERSION   REPLICAS   READY   UPDATED   UNAVAILABLE
kubeadmcontrolplane.controlplane.cluster.x-k8s.io/clusterclass-quickstart-XXXX    25m   true          true                   v1.22.0   1          1       1

NAME                                                                                    CLUSTER                   AGE   PROVIDERID                                                                PHASE     VERSION
machine.cluster.x-k8s.io/clusterclass-quickstart-linux-workers-XXXX                     clusterclass-quickstart   10m   docker:////clusterclass-quickstart-linux-workers-XXXX                     Running   v1.22.0
machine.cluster.x-k8s.io/clusterclass-quickstart-XXXX                                   clusterclass-quickstart   13m   docker:////clusterclass-quickstart-skdmm-XXXX                             Running   v1.22.0
```

### Clean Up

Delete workload cluster.
```bash

> kubectl delete cluster capi-quickstart

Reviewer 2: Note: this will only work if you rename the cluster (as suggested) to `capi-quickstart`.

```
<aside class="note warning">

IMPORTANT: In order to ensure a proper cleanup of your infrastructure you must always delete the cluster object. Deleting the entire cluster template with `kubectl delete -f capi-quickstart.yaml` might lead to pending resources to be cleaned up manually.
</aside>

Delete management cluster
```bash

> kind delete cluster

Reviewer 2: I don't think the name matches the kind cluster we created above (also: we probably should refer only to the hack script **or** to the quickstart guide, to avoid variants in the kind cluster name)

Author: I'll update to the name from the hack script.

Reviewer 2: One note: using the hack script requires cloning the repo. Always nice if that's not necessary 
(Or I guess we can also do `curl | sh`)

Author: I'm not sure about curl | sh, but I'll link to the raw script so that users can download and run as they wish.

Reviewer 2: Yeah I don't really like `curl | sh`. Just not great if a user has to clone the cluster-api repo to run a quickstart.

            eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'
```

> The above file can be found in the [CAPI repo](../../../../../test/infrastructure/docker/examples/clusterclass.yaml)

Reviewer: I'm not sure relative paths works when the site is published (and the relative path is outside the book)
